pkgs/misc/emulators/dosbox-staging/default.nix

```
@@ -0,0 +1,39 @@
+{ stdenv, lib, fetchFromGitHub, autoconf, autoconf-archive, automake, pkgconfig, alsaLib, SDL2, SDL2_net, libGL, libGLU, libogg, opusfile }:
+
+stdenv.mkDerivation rec {
+  pname = "dosbox-staging";
+  version = "v0.75.0-rc1";
+
+  src = fetchFromGitHub {
+    owner = "dreamer";
+    repo = pname;
+    rev = version;
+    sha256 = "0yr4ifpwk49cgy60y56hkpsyx89f8n45yqq5a9f8rvyq56m1nr1x";
+  };
+
+  hardeningDisable = [ "format" ];
+
+  nativeBuildInputs = [ pkgconfig ];
+  buildInputs = [ autoconf autoconf-archive automake SDL2 libGL libGLU libogg opusfile alsaLib SDL2_net ];
+
+  preConfigure = "./autogen.sh";
+
+  preBuild = ''
+    buildFlagsArray=( "CXXFLAGS=-O3 -DNDEBUG" )
+  '';
+
+  postInstall = ''
+    mkdir -p $out/share/applications
+    cp $src/contrib/linux/dosbox-staging.desktop $out/share/applications/dosbox-staging.desktop
+    mkdir -p $out/share/icons/hicolor/scalable/apps
+    cp $src/contrib/icons/dosbox-staging.svg $out/share/icons/hicolor/scalable/apps/dosbox-staging.svg
+  '';
+
+  meta = with stdenv.lib; {
+    description = "A modernized DOS emulator";
+    homepage = "https://github.com/dreamer/dosbox-staging";
+    license = licenses.gpl2;
+    maintainers = with maintainers; [ joshuafern ];
+    platforms = platforms.unix;
```

One thing I would suggest is to make it so this derivation can be installed alongside the other dosbox derivation. Right now they conflict on `bin/dosbox`. You could move it to `bin/dosbox-staging`, patch as needed, and then symlink `dosbox` in.

To let dosbox override the symlink:
```suggestion
    platforms = platforms.unix;
    priority = 10;
```

---

(dosbox-staging maintainer here, I'm not NixOS user but want to make the packaging experience better and care about end-user experience)

@lourkeur After this change, is it going to be clear to the end users, that dosbox-staging is supposed to be invoked via `dosbox` command?

We have technical reasons for keeping `dosbox` program name, that's why we don't generate `dosbox-staging` binary by default. Program is supposed to be drop-in replacement for upstream DOSBox. This affects also:

- man page (also named `dosbox` to match program name)
- XDG config location (`$XDG_CONFIG_HOME/dosbox/`)

It does not affect icons (`dosbox-staging.{png,svg}`).

---

@dreamer Thanks a lot for coming. I'm not very familiar with the DOSBox family so your input is very welcome.

In my vision for this derivation, `dosbox-staging` is the canonical command to invoke dosbox-staging and `dosbox` is aliased to it if possible. In case both derivations are deployed, upstream takes `dosbox`, but dosbox-staging is still accessible. That being said, that might be excessive patching on my part.

Would it make sense at all for someone to want to have both DOSBox and dosbox-staging side-by-side? If not, then I'll retract that. Can they share `$XDG_CONFIG_HOME/dosbox` or do they clash?

---

> In my vision for this derivation, `dosbox-staging` is the canonical command to invoke dosbox-staging and `dosbox` is aliased to it if possible. In case both derivations are deployed, upstream takes `dosbox`, but dosbox-staging is still accessible. That being said, that might be excessive patching on my part.

Well, perhaps the solution you suggested is the best one and I need to adjust my builds somehow, let's find out :)

How NixOS handles `wine` and `wine-staging`? If there's already established pattern for that, then maybe following the same scheme would be the most intuitive for users?

Our situation is a bit different - we started as staging repo, ended up a fork - but from user perspective, it might be a meaningless distinction.

> Can they share `$XDG_CONFIG_HOME/dosbox` or do they clash?

`$XDG_CONFIG_HOME/dosbox/` does not clash… mostly because upstream does not (will not) respect XDG. In an unlikely possibility, that they will change their minds - actual default config file is called differently (`$XDG_CONFIG_HOME/dosbox/dosbox-staging.conf`).

> Would it make sense at all for someone to want to have both DOSBox and dosbox-staging side-by-side? If not, then I'll retract that.

Unfortunately, yes - people like me (dosbox developers, testers). I presume (perhaps wrongly), that most users just want 1 dosbox installation, that works well for their games.

---

### Current

Let's first go over the current behavior, it works like this: If a user installs both dosbox and dosbox-staging declaratively they'll get a warning like this:
```
collision between `/nix/store/d8v98l381qrqy5ls9xkccpx47w1n1y2k-dosbox-0.74-3/bin/dosbox' and `/nix/store/6zma6j90lwfc84fs3130z6b0l7dyk2yc-dosbox-staging-v0.75.0-rc1/bin/dosbox'
```
Then they can fix their configuration to pick one or the other. This doesn't mean it's impossible to run both at the same time, you could simply run `nix-shell -p dosbox`
to open a shell with vanilla dosbox, or any package, at any time.

Also note that this is pretty common in nixpkgs. There's a ridiculous amount of software packaged and files are bound to collide. Of course that doesn't mean we should settle with the status quo!

### Priority

On the priority suggestion, let's look at the documentation about how it works exactly.
From `nix-env --help`:
> priority can be changed to resolve filename clashes. The user environment build script uses the meta.priority attribute of derivations to resolve filename collisions between packages. Lower priority values denote a higher priority. For instance, the GCC wrapper package and the Binutils package in Nixpkgs both have a file bin/ld, so previously if you tried to install both you would get a collision. Now, on the other hand, the GCC wrapper declares a higher priority than Binutils, so the former’s bin/ld is symlinked in the user environment.

Thinking ahead a bit, what happens when we package other forks, like dosbox-x?
If dosbox-x and dosbox-staging both have symlinks to /bin/dosbox and both are priority 10 which will take precedence? Do we have to make a judgement call here and say that one project is more important by changing the priority to be higher?

---

@dreamer wine-staging uses `bin/wine`, good point.

@joshuafern That's all correct. I think my concern is not worth the cost given what dreamer said so I hereby retract that suggestion.

---

>I think my concern is not worth the cost given what dreamer said so I hereby retract that suggestion.
It's an interesting suggestion, food for thought :)